sys/net/network_layer/fib/fib.c

```diff
@@ -609,6 +691,712 @@ int fib_get_num_used_entries(void)
+    if ((fib_sr == NULL) || (sr_lifetime == 0)) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    for (size_t i = 0; i < FIB_MAX_FIB_SR; ++i) {
+        if (fib_sr_table[i].sr_lifetime.seconds == 0
+            && fib_sr_table[i].sr_lifetime.microseconds == 0) {
+            fib_sr_table[i].sr_iface_id = sr_iface_id;
+            fib_sr_table[i].sr_flags = sr_flags;
+            fib_sr_table[i].sr_path = NULL;
+            fib_sr_table[i].sr_dest = NULL;
+            if (sr_lifetime < FIB_LIFETIME_NO_EXPIRE) {
+                fib_ms_to_timex(sr_lifetime, &fib_sr_table[i].sr_lifetime);
+            }
+            else {
+                fib_sr_table[i].sr_lifetime.seconds = FIB_LIFETIME_NO_EXPIRE;
+                fib_sr_table[i].sr_lifetime.microseconds = FIB_LIFETIME_NO_EXPIRE;
+            }
+            *fib_sr = &fib_sr_table[i];
+            mutex_unlock(&mtx_access);
+            return 0;
+        }
+    }
+
+    mutex_unlock(&mtx_access);
+    return -ENOBUFS;
+}
+
+
+static int fib_sr_check_lifetime(fib_sr_t *fib_sr, timex_t *now)
+{
+    /* check if the lifetime expired */
+    if (timex_cmp(*now, fib_sr->sr_lifetime) > -1) {
+        /* remove this sr if its lifetime expired */
+        fib_sr->sr_lifetime.seconds = 0;
+        fib_sr->sr_lifetime.microseconds = 0;
+
+        if (fib_sr->sr_path != NULL) {
+            fib_sr_entry_t *elt;
+            LL_FOREACH(fib_sr->sr_path, elt) universal_address_rem(elt->address);
+            fib_sr->sr_path = NULL;
+        }
+
+        /* Now we set the passed `fib_sr` to NULL and return an errorcode */
+        fib_sr = NULL;
+        return -ENOENT;
+    }
+    return 0;
+}
+
+
+static int fib_sr_new_entry(uint8_t *addr, size_t addr_size, fib_sr_entry_t **new_entry)
+{
+    for (size_t i = 0; i < FIB_MAX_FIB_SR_ENTRIES; ++i) {
+        if (fib_sr_entry_table[i].address == NULL) {
+            fib_sr_entry_table[i].address = universal_address_add(addr, addr_size);
+            if (fib_sr_entry_table[i].address == NULL) {
+                return -ENOMEM;
+            }
+            else {
+                (void)new_entry;
+                *new_entry = &fib_sr_entry_table[i];
+                return 0;
+            }
+        }
+    }
+    return -ENOMEM;
+}
+
+int fib_sr_read_head(fib_sr_t *fib_sr, kernel_pid_t *iface_id, uint32_t *sr_flags,
+                     uint32_t *sr_lifetime)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+    if (fib_sr == NULL || iface_id == NULL || sr_flags == NULL || sr_lifetime == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr_check_lifetime(fib_sr, &now) == -ENOENT) {
+        mutex_unlock(&mtx_access);
+        return -ENOENT;
+    }
+
+    *iface_id = fib_sr->sr_iface_id;
+    *sr_flags = fib_sr->sr_flags;
+    timex_t tm = timex_sub(fib_sr->sr_lifetime, now);
+    *sr_lifetime = (tm.seconds * 1000) + (tm.microseconds / 1000);
```

Comment:
you could also use the `timex_uint64()` function, but would require to to change the type of `sr_lifetime` to `uint64_t`